manager/integration/tests/test_node.py

```diff
@@ -0,0 +1,359 @@
+            assert not disk["allowScheduling"]
+            assert disk["storageReserved"] == Gi
+            assert disk["storageScheduled"] == 0
+
+    # delete disk
+    update_disk = generate_update_disk(disks)
+    delete_disk = []
+    for disk in update_disk:
+        # keep default disk for other tests
+        if disk["path"] != disk_path1 and disk["path"] != disk_path2:
+            delete_disk.append(disk)
+    node = node.diskUpdate(disks=delete_disk)
+    assert len(node["disks"]) == len(delete_disk)
+    # clear disks
+    clean_disk(client, 'vol-disk-1', 'vol-disk-2')
+
+
+def create_volume(client, vol_name, size, node_id, r_num):
+    volume = client.create_volume(name=vol_name, size=size,
+                                  numberOfReplicas=r_num)
+    assert volume["numberOfReplicas"] == r_num
+    assert volume["frontend"] == "blockdev"
+
+    volume = common.wait_for_volume_detached(client, vol_name)
+    assert len(volume["replicas"]) == r_num
+
+    assert volume["state"] == "detached"
+    assert volume["created"] != ""
+
+    volumeByName = client.by_id_volume(vol_name)
+    assert volumeByName["name"] == volume["name"]
+    assert volumeByName["size"] == volume["size"]
+    assert volumeByName["numberOfReplicas"] == volume["numberOfReplicas"]
+    assert volumeByName["state"] == volume["state"]
+    assert volumeByName["created"] == volume["created"]
+
+    volume.attach(hostId=node_id)
+    volume = common.wait_for_volume_healthy(client, vol_name)
+
+    return volume
+
+
+def clean_volume(client, vol_name):
+    volume = client.by_id_volume(vol_name)
+    volume.detach()
+    common.wait_for_volume_detached(client, vol_name)
+    client.delete(volume)
+    common.wait_for_volume_delete(client, vol_name)
+
+
+def test_replica_scheduler(clients):  # NOQA
+    # get a random client
+    for host_id, client in clients.iteritems():
+        break
+
+    nodes = client.list_node()
+    # delete all disks on each node
+    for node in nodes:
+        disks = node["disks"]
+        # set allowScheduling to false
+        for fsid, disk in disks.iteritems():
+            disk["allowScheduling"] = False
+        update_disks = generate_update_disk(disks)
+        node = node.diskUpdate(disks=update_disks)
+        node = client.by_id_node(node["name"])
+        for fsid, disk in node["disks"].iteritems():
+            assert not disk["allowScheduling"]
+
+        # wait for node controller update storageScheduled value
+        time.sleep(30)
+        node = node.diskUpdate(disks=[])
+        assert len(node["disks"]) == 0
+
+    # test there's no disk fit for volume
+    vol_name = common.generate_volume_name()
+    volume = client.create_volume(name=vol_name,
+                                  size=SIZE, numberOfReplicas=len(nodes))
+    volume = common.wait_for_volume_status(client, vol_name,
+                                           common.VOLUME_FIELD_STATE, "")
+    assert volume["state"] == ""
```

Comment:
This is not enough to determine if the volume is un-schedulable. When the volume was created its state can be "" as well. We need to figure out a way to report un-schedulable volumes.

Reply:
I will add a new `Conditions` field in the volume to address this issue.

Reply:
@JacieChao I've added Conditions to volumeStatus, and make scheduled as one condition.

See this commit for details on how it works: https://github.com/rancher/longhorn-manager/commit/e9ba26bad604dcb9c492c4790a02adbe02f3aa49

I've exposed in the API as volume.conditions, like this:
```
"conditions": {
"scheduled": {
"lastProbeTime": "2018-07-20T02:07:50Z",
"lastTransitionTime": "2018-07-20T02:07:37Z",
"message": "",
"reason": "ReplicaSchedulingFailure",
"status": "False",
"type": "scheduled",
},
},
```

`status: False` (using k8s words) means scheduling failed, as stated in the reason.